### PR TITLE
Orderability Semantics

### PR DESCRIPTION
**Pull request by mikepersonick**
Implementation of total orderabillity semantics including across types.

https://issues.apache.org/jira/browse/TINKERPOP-2641

https://tinkerpop.apache.org/docs/3.6.0-SNAPSHOT/dev/provider/#_composite_types_3

### CODE REVIEW

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/OrderabilityTest.java [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+        checkOrderedResults(Arrays.asList(
+                convertToEdge("josh", "created", "lop"),        // eid = 11
+                convertToEdge("josh", "created", "ripple")      // eid = 10
+        ), traversal);
+    }
+
+    /**
+     * Order desc by vertex and then vertex property id in path.
+     *
+     * Note to graph providers: if your graph does not support user-assigned vertex ids and vertex property ids you
```
Comment (divijvaidya) [inline gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/OrderabilityTest.java:402]: you can already check the feature support in the tests using `@FeatureRequirement` annotation.

```
@FeatureRequirement(featureClass = ElementFeatures.class, feature = FEATURE_USER_SUPPLIED_IDS)
```

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/OrderabilityComparator.java [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+            Comparator.<Map.Entry,Object>comparing(Map.Entry::getKey, INSTANCE).thenComparing(Map.Entry::getValue, INSTANCE);
+
+    /**
+     * Sort unknown types first by classname, then by natural toString().
+     */
+    private static final Comparator<Object> unknownTypeComparator =
+            Comparator.comparing(f -> f.getClass().getName()).thenComparing(Object::toString);
+
+    /**
+     * The comparators that will be used for the types defined above in the {@link #types} array. Be sure to keep them
```
Comment (divijvaidya) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/OrderabilityComparator.java:161]: optional suggestion:
We can avoid potential bugs due to out of sync arrays by storing the mapping btw types and comparators in a LinkedHashMap<Class, Comparator>.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/OrderabilityComparator.java [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+     * Return the first index of the Class array for which the supplied object is an instance. Returns
+     * {@link #UNKNOWN_TYPE} if not an instance of any of the supplied classes.
+     */
+    private static final int type(final Class[] types, final Object o) {
+        for (int i = 0; i < types.length; i++) {
+            if (types[i].isInstance(o)) {
+                return i;
+            }
+        }
+        return UNKNOWN_TYPE;
```
Comment (divijvaidya) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/OrderabilityComparator.java:197]: What are the cases where we legitimately expect an unknown type? It should probably be an exception if we are dealing with a type not defined in the semantics.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/OrderabilityComparator.java [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+ *
+ * @author Mike Personick
+ */
+public class OrderabilityComparator implements Comparator<Object> {
+
+    public static final Comparator<Object> INSTANCE = Comparator.comparing(
+            // first transform (strip Traverser layer)
+            OrderabilityComparator::transform,
+                    // then handle nulls
+                    Comparator.nullsFirst(
```
Comment (divijvaidya) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/OrderabilityComparator.java:51]: nulls go last for desc. Please add a unit test if not already present.

**Review by divijvaidya — CHANGES_REQUESTED** [review]

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/OrderabilityComparator.java (hunk 4 above)
Comment (mikepersonick) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/OrderabilityComparator.java:51]: It is handled in Order.desc. Added a test.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/OrderabilityComparator.java (hunk 3 above)
Comment (mikepersonick) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/OrderabilityComparator.java:197]: g.inject on a local tinkergraph? lambdas? maybe others. Probably would never happen over a remote connection. Part of Orderability semantics is no exceptions thrown - total universal order. I will update the semantics documentation if it's not in there already.

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/OrderabilityTest.java (hunk 1 above)
Comment (mikepersonick) [reply gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/OrderabilityTest.java]: done

File: gremlin-test/features/semantics/Orderability.feature [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+      | java |
+      | java |
+      | josh |
+      | lop  |
+      | marko |
+      | peter |
+      | ripple |
+      | vadas  |
+
+# This exposes an underlying problem where .value() does not preserve the order (should it?). This will be
```
Comment (mikepersonick) [inline gremlin-test/features/semantics/Orderability.feature:43]: Remove this

File: gremlin-test/features/semantics/Orderability.feature [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+      """
+    When iterated to list
+    Then the result should be ordered
+      | result |
+      | java |
+      | ripple |
+      | java |
+      | lop |
+
+  Scenario: g_V_out_out_values_asXheadX_path_order_byXascX_selectXheadX
```
Comment (mikepersonick) [inline gremlin-test/features/semantics/Orderability.feature:328]: "@UserSuppliedVertexIds"

File: gremlin-dotnet/test/Gremlin.Net.IntegrationTest/Gherkin/GherkinTestRunner.cs [hunk 7]
```diff
@@ -85,6 +85,7 @@ public class GherkinTestRunner
                     "g_V_hasXperson_name_markoX_bothXknowsX_groupCount_byXvaluesXnameX_foldX",
                     IgnoreReason.ArrayKeysInMapNotAssertingInGherkin
                 },
+                {"g_V_properties_order_id", IgnoreReason.NoReason},
```
Comment (mikepersonick) [inline gremlin-dotnet/test/Gremlin.Net.IntegrationTest/Gherkin/GherkinTestRunner.cs:88]: @spmallette

**Pull request closed without merging**